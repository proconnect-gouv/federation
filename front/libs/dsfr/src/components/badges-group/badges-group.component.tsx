import React from 'react';

import type { BadgeInterface } from '../../interfaces';
import { BadgeComponent } from '../badge';

interface BadgesGroupComponentProps {
  item: BadgeInterface[];
}

export const BadgesGroupComponent = React.memo(({ item }: BadgesGroupComponentProps) => (
  <div className="fr-badge-group">
    {item.map((badge, index) => {
      const uniqkey = `badge-group::badge::${index}`;
      return <BadgeComponent key={uniqkey} colorName={badge.colorName} label={badge.label} />;
    })}
  </div>
));

BadgesGroupComponent.displayName = 'BadgesGroupComponent';
