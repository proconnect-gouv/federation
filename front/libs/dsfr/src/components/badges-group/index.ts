export * from './badges-group.component';
