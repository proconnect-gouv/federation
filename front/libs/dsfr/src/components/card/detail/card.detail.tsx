import classnames from 'classnames';
import React from 'react';

interface CardDetailComponentProps {
  content: string;
  dataTestId?: string | undefined;
  className?: string | undefined;
}

export const CardDetailComponent = React.memo(
  ({ className = undefined, content, dataTestId = undefined }: CardDetailComponentProps) => (
    <p className={classnames('fr-card__detail', className)} data-testid={dataTestId}>
      {content}
    </p>
  ),
);

CardDetailComponent.displayName = 'CardDetailComponent';
