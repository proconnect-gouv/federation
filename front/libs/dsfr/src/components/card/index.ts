/* istanbul ignore file */

// declarative file
export * from './card.component';
