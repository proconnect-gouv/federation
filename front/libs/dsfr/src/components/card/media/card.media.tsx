import React from 'react';

import type { MediaInterface } from '../../../interfaces';

export const CardMediaComponent = React.memo(({ alt = '', src }: MediaInterface) => (
  <div className="fr-card__img">
    <img alt={alt} className="fr-responsive-img" src={src} />
  </div>
));

CardMediaComponent.displayName = 'CardMediaComponent';
