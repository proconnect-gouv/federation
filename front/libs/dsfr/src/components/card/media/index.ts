/* istanbul ignore file */

// declarative file
export * from './card.media';
