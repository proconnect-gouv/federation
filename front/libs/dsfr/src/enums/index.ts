/* istanbul ignore file */

// declarative file
export * from './alert-types.enum';
export * from './button-types.enum';
export * from './icon-placement.enum';
export * from './pagination.enum';
export * from './priorities.enum';
export * from './sizes.enum';
