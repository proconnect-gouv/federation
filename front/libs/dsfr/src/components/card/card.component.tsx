import classnames from 'classnames';
import type { PropsWithChildren } from 'react';
import React from 'react';
import type { To } from 'react-router-dom';
import { Link } from 'react-router-dom';

import { HeadingTag } from '@fc/common';

import { CardBackgrounds, Sizes } from '../../enums';
import type { BadgeInterface, MediaInterface } from '../../interfaces';
import { BadgesGroupComponent } from '../badges-group';
import { CardDetailComponent } from './detail';
import { CardMediaComponent } from './media/card.media';

interface CardComponentDetail {
  // @TODO
  // use a namespace interface
  // to not have a duplicate interface between detail and this one
  content: string | undefined;
  className?: string | undefined;
}

interface CardComponentDetails {
  top?: CardComponentDetail | undefined;
  bottom?: CardComponentDetail | undefined;
}

interface CardComponentProps extends PropsWithChildren {
  title: string;
  enlargeLink?: boolean;
  size?: Sizes;
  media?: MediaInterface;
  background?: CardBackgrounds | undefined;
  details?: CardComponentDetails | undefined;
  isHorizontal?: boolean;
  Heading?: HeadingTag;
  link?: To | undefined;
  badges?: BadgeInterface[] | undefined;
  className?: string | undefined;
}

export const CardComponent = React.memo(
  ({
    Heading = HeadingTag.H3,
    background = CardBackgrounds.NO_BACKGROUND,
    badges,
    children: description,
    className,
    details,
    enlargeLink = false,
    isHorizontal = false,
    link = '#',
    media,
    size = Sizes.MEDIUM,
    title,
  }: CardComponentProps) => (
    <div
      className={classnames(
        `fr-card fr-card--${size} fr-card--${background}`,
        {
          // Class CSS
          // eslint-disable-next-line @typescript-eslint/naming-convention
          'fr-card--horizontal': isHorizontal,
          // Class CSS
          // eslint-disable-next-line @typescript-eslint/naming-convention
          'fr-enlarge-link': enlargeLink,
        },
        className,
      )}
      data-testid="CardComponent">
      <div className="fr-card__body">
        <div className="fr-card__content">
          <Heading className="fr-card__title" data-testid="CardComponent-title">
            <Link to={link}>{title}</Link>
          </Heading>
          <p className="fr-card__desc">{description}</p>
          {details?.bottom && details?.bottom.content && (
            <div className="fr-card__end">
              <CardDetailComponent
                className={details.bottom.className}
                content={details.bottom.content}
                dataTestId="CardComponent-detail-bottom"
              />
            </div>
          )}
          <div className="fr-card__start">
            {media && <CardMediaComponent alt={media.alt} src={media.src} />}
            {badges && <BadgesGroupComponent item={badges} />}
            {details?.top && details?.top.content && (
              <CardDetailComponent
                className={details.top.className}
                content={details.top.content}
                dataTestId="CardComponent-detail-top"
              />
            )}
          </div>
        </div>
      </div>
    </div>
  ),
);

CardComponent.displayName = 'CardComponent';
