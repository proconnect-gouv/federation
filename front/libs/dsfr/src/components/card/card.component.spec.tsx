import { render } from '@testing-library/react';
import { Link } from 'react-router-dom';

import { HeadingTag } from '@fc/common';

import { CardBackgrounds, Sizes } from '../../enums';
import { BadgesGroupComponent } from '../badges-group';
import { CardComponent } from './card.component';
import { CardDetailComponent } from './detail';
import { CardMediaComponent } from './media';

jest.mock('./media/card.media');
jest.mock('./detail/card.detail');
jest.mock('../badges-group/badges-group.component');

describe('CardComponent', () => {
  it('should match the snapshot, with default values', () => {
    // When
    const { container } = render(
      <CardComponent title="Card title mock">any description text treat as children</CardComponent>,
    );

    // Then
    expect(container).toMatchSnapshot();
  });

  it('should match the snapshot, with all optionnal values', () => {
    // When
    const { container } = render(
      <CardComponent
        enlargeLink
        isHorizontal
        background={CardBackgrounds.SHADOW}
        badges={[{ label: 'any-badge' }]}
        className="any-classname-mock"
        details={{
          bottom: { className: 'any-classname-bottom-mock', content: 'any-content-bottom-mock' },
          top: { className: 'any-classname-top-mock', content: 'any-content-top-mock' },
        }}
        Heading={HeadingTag.H1}
        link="any-link-mock"
        media={{ alt: 'media-alt-mock', src: 'media-src-mock' }}
        size={Sizes.LARGE}
        title="Card title mock">
        any description text treat as children
      </CardComponent>,
    );

    // Then
    expect(container).toMatchSnapshot();
  });

  it('should optionnal classnames on the main container', () => {
    // When
    const { container } = render(
      <CardComponent
        enlargeLink
        isHorizontal
        background={CardBackgrounds.SHADOW}
        className="any-classname-mock"
        size={Sizes.LARGE}
        title="Card title mock">
        any description text treat as children
      </CardComponent>,
    );
    const element = container.firstChild;

    // Then
    expect(element).toHaveClass('fr-enlarge-link');
    expect(element).toHaveClass('fr-card--horizontal');
    expect(element).toHaveClass('fr-card--shadow');
    expect(element).toHaveClass('any-classname-mock');
    expect(element).toHaveClass('fr-card--lg');
  });

  it('should render the heading tag with a link and the title', () => {
    // When
    const { getByRole } = render(
      <CardComponent Heading={HeadingTag.H1} link="any-link-mock" title="Card title mock">
        any description text treat as children
      </CardComponent>,
    );
    const titleElt = getByRole('heading', { level: 1 });

    // Then
    expect(titleElt).toBeInTheDocument();
    expect(titleElt).toHaveClass('fr-card__title');
    expect(titleElt).toHaveAttribute('data-testid', 'CardComponent-title');
    expect(Link).toHaveBeenCalledOnce();
    expect(Link).toHaveBeenCalledWith(
      {
        children: 'Card title mock',
        to: 'any-link-mock',
      },
      {},
    );
  });

  it('should render the description', () => {
    // When
    const { getByText } = render(
      <CardComponent title="Card title mock">any description text treat as children</CardComponent>,
    );
    const element = getByText('any description text treat as children');

    // Then
    expect(element).toBeInTheDocument();
  });

  it('should render the card details (top, bottom)', () => {
    // When
    render(
      <CardComponent
        details={{
          bottom: { className: 'any-classname-bottom-mock', content: 'any-content-bottom-mock' },
          top: { className: 'any-classname-top-mock', content: 'any-content-top-mock' },
        }}
        title="Card title mock">
        any description text treat as children
      </CardComponent>,
    );

    // Then
    expect(CardDetailComponent).toHaveBeenCalledTimes(2);
    expect(CardDetailComponent).toHaveBeenNthCalledWith(
      1,
      {
        className: 'any-classname-bottom-mock',
        content: 'any-content-bottom-mock',
        dataTestId: 'CardComponent-detail-bottom',
      },
      {},
    );
    expect(CardDetailComponent).toHaveBeenNthCalledWith(
      2,
      {
        className: 'any-classname-top-mock',
        content: 'any-content-top-mock',
        dataTestId: 'CardComponent-detail-top',
      },
      {},
    );
  });

  it('should render the media', () => {
    // When
    render(
      <CardComponent
        media={{ alt: 'media-alt-mock', src: 'media-src-mock' }}
        title="Card title mock">
        any description text treat as children
      </CardComponent>,
    );

    // Then
    expect(CardMediaComponent).toHaveBeenCalledOnce();
    expect(CardMediaComponent).toHaveBeenCalledWith(
      {
        alt: 'media-alt-mock',
        src: 'media-src-mock',
      },
      {},
    );
  });

  it('should render the badges group', () => {
    // When
    render(
      <CardComponent badges={[{ label: 'any-badge-mock' }]} title="Card title mock">
        any description text treat as children
      </CardComponent>,
    );

    // Then
    expect(BadgesGroupComponent).toHaveBeenCalledOnce();
    expect(BadgesGroupComponent).toHaveBeenCalledWith(
      {
        item: [{ label: 'any-badge-mock' }],
      },
      {},
    );
  });
});
