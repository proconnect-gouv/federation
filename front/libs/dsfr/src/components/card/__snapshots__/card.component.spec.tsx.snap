// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardComponent should match the snapshot, with all optionnal values 1`] = `
<div>
  <div
    class="fr-card fr-card--lg fr-card--shadow fr-card--horizontal fr-enlarge-link any-classname-mock"
    data-testid="CardComponent"
  >
    <div
      class="fr-card__body"
    >
      <div
        class="fr-card__content"
      >
        <h1
          class="fr-card__title"
          data-testid="CardComponent-title"
        >
          <span
            data-mockid="Link"
          >
            Card title mock
          </span>
        </h1>
        <p
          class="fr-card__desc"
        >
          any description text treat as children
        </p>
        <div
          class="fr-card__end"
        >
          <div>
            CardDetailComponent
          </div>
        </div>
        <div
          class="fr-card__start"
        >
          <div>
            CardMediaComponent
          </div>
          <div>
            BadgeGroupComponent
          </div>
          <div>
            CardDetailComponent
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CardComponent should match the snapshot, with default values 1`] = `
<div>
  <div
    class="fr-card fr-card--md fr-card--no-background"
    data-testid="CardComponent"
  >
    <div
      class="fr-card__body"
    >
      <div
        class="fr-card__content"
      >
        <h3
          class="fr-card__title"
          data-testid="CardComponent-title"
        >
          <span
            data-mockid="Link"
          >
            Card title mock
          </span>
        </h3>
        <p
          class="fr-card__desc"
        >
          any description text treat as children
        </p>
        <div
          class="fr-card__start"
        />
      </div>
    </div>
  </div>
</div>
`;
